src/Microsoft.ApiDesignGuidelines.Analyzers/Core/NestedTypesShouldNotBeVisible.cs

```diff
@@ -19,31 +21,112 @@ public abstract class NestedTypesShouldNotBeVisibleAnalyzer : DiagnosticAnalyzer
-                                                                             isEnabledByDefault: false,
+                                                                             Category,
+                                                                             Severity,
+                                                                             IsEnabledByDefault,
                                                                              description: s_localizableDescription,
-                                                                             helpLinkUri: null,     // TODO: add MSDN url
-                                                                             customTags: WellKnownDiagnosticTags.Telemetry);
+                                                                             helpLinkUri: HelpLinkUrl,
+                                                                             customTags: CustomTags);
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DefaultRule, VisualBasicModuleRule);
 
         public override void Initialize(AnalysisContext analysisContext)
         {
-            
+            analysisContext.RegisterCompilationStartAction(
+                compilationStartContext =>
+                {
+                    var compilation = compilationStartContext.Compilation;
+
+                    INamedTypeSymbol enumeratorType = compilation.GetTypeByMetadataName("System.Collections.IEnumerator");
+                    INamedTypeSymbol dataSetType = compilation.GetTypeByMetadataName("System.Data.DataSet");
+                    INamedTypeSymbol dataTableType  = compilation.GetTypeByMetadataName("System.Data.DataTable");
+                    INamedTypeSymbol dataRowType = compilation.GetTypeByMetadataName("System.Data.DataRow");
+
+                    compilationStartContext.RegisterSymbolAction(
+                        symbolAnalysisContext =>
+                        {
+                            var nestedType = (INamedTypeSymbol)symbolAnalysisContext.Symbol;
+                            INamedTypeSymbol containingType = nestedType.ContainingType;
+                            if (containingType == null)
+                            {
+                                return;
+                            }
+
+                            // The Framework Design Guidelines (see 4.9 Nested Types) say that it is okay
+                            // to expose nested types for advanced customization and subclassing scenarios,
+                            // so, following FxCop's implementation of this rule, we allow protected and
+                            // internal nested types.
+                            if (nestedType.DeclaredAccessibility != Accessibility.Public)
+                            {
+                                return;
+                            }
+
```

Comment:
Interesting. @lgolding  you can check that with `nestedType.GetResultantVisibility() == SymbolVisibility.Public`